include/eve/detail/function/simd/x86/swizzle.hpp

```diff
@@ -0,0 +1,258 @@
+  // shuffle requires cross-api re-targeting (ie AVX512 128bits calling back SSSE3)
+  // so instead of our classical one-size-fit-all overloads, we need to split it into
+  // API supports overloads
+  //================================================================================================
+
+  //================================================================================================
+  // Unary swizzle - logical on AVX512 ABI, call general case on others
+  //================================================================================================
+  template<typename T, typename N, x86_abi ABI, shuffle_pattern Pattern>
+  EVE_FORCEINLINE auto basic_swizzle_ ( EVE_SUPPORTS(sse2_)
+                                      , logical<wide<T,N,ABI>> const& v, Pattern p
+                                      ) noexcept
+  {
+    if constexpr( current_api >= avx512 ) return to_logical((v.mask())[p]);
+    else                                  return basic_swizzle_(EVE_RETARGET(cpu_),v,p);
+  }
+
+  //================================================================================================
+  // Check a pattern fits the half-lane rules for x86 pseudo-shuffle
+  //================================================================================================
+  template<std::ptrdiff_t... I>
+  consteval bool is_x86_shuffle_compatible(pattern_t<I...> p)
+  {
+    constexpr std::ptrdiff_t c  = sizeof...(I);
+    std::ptrdiff_t idx[c];
+
+    for(int i=0;i<c  ;++i)  idx[i] = p(i,c);
+    for(int i=0;i<c/2;++i)  if(idx[i] >= c/2)                 return false;
+    for(int i=c/2;i<c;++i)  if(idx[i] <  c/2 && idx[i] != -1) return false;
+
+    return true;
+  }
+
+  //================================================================================================
+  // SSE2-SSSE3 variant
+  //================================================================================================
+  template<typename T, typename N, x86_abi ABI, shuffle_pattern Pattern>
+  EVE_FORCEINLINE auto basic_swizzle_( EVE_SUPPORTS(sse2_), wide<T,N,ABI> const& v, Pattern const&)
+  {
+    constexpr auto sz = Pattern::size(N::value);
+    using that_t      = as_wide_t<wide<T,N,ABI>,fixed<sz>>;
+
+    constexpr Pattern q = {};
+
+    // We're swizzling so much we aggregate the output
+    if constexpr( has_aggregated_abi_v<that_t> )
+    {
+      return aggregate_swizzle(v,q);
+    }
+    else if constexpr( current_api >= ssse3 )
+    {
+      using st_t    = typename that_t::storage_type;
+      using bytes_t = typename that_t::template rebind<std::uint8_t,fixed<16>>;
+      using i_t     = as_integer_t<wide<T,N,ABI>>;
+
+      return that_t ( (st_t)_mm_shuffle_epi8( bit_cast(v,as_<i_t>{}).storage()
+                                            , as_bytes<that_t>(q,as_<bytes_t>())
+                                            )
+                    );
```

Comment:
And zeroes?

Reply:
zeroes are handled directly by shuffle_epi8

From the Intel guide
```
__m128i _mm_shuffle_epi8 (__m128i a, __m128i b)
Synopsis
__m128i _mm_shuffle_epi8 (__m128i a, __m128i b)
#include <tmmintrin.h>
Instruction: pshufb xmm, xmm
CPUID Flags: SSSE3
Description
Shuffle packed 8-bit integers in a according to shuffle control mask in the corresponding 8-bit element of b, and store the results in dst.
Operation
FOR j := 0 to 15
	i := j*8
	IF b[i+7] == 1
		dst[i+7:i] := 0
	ELSE
		index[3:0] := b[i+3:i]
		dst[i+7:i] := a[index*8+7:index*8]
	FI
ENDFOR
```